File: Daily/200.number-of-islands.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int numIslands(vector<vector<char>>& grid) {
    int islandNumber = 0;
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid[0].size(); j++) {
        if (grid[i][j] == LAND) {
          islandNumber++;
          flood(grid, i, j);
        }
      }
    }
    return islandNumber;
  }

 private:
  const char LAND = '1';
  const char WATER = '0';

  inline void flood(vector<vector<char>>& grid, int i, int j) {
    if ((i < 0 || i >= grid.size()) || (j < 0 || j >= grid[0].size()) ||
        grid[i][j] == WATER) {
      return;
    }
    grid[i][j] = WATER;
    flood(grid, i + 1, j);
    flood(grid, i - 1, j);
    flood(grid, i, j + 1);
    flood(grid, i, j - 1);
  }
};
// @lc code=end

File: job-2022-autumn/200.number-of-islands.js
/*
 * @lc app=leetcode id=200 lang=javascript
 *
 * [200] Number of Islands
 */

// @lc code=start
/**
 * @param {string[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  const LAND = '1';
  const WATER = '0';
  const m = grid.length;
  const n = grid[0].length;

  /**
   * @param {number} i
   * @param {number} j
   */
  const flood = (i, j) => {
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] === WATER) {
      return;
    } else if (grid[i][j] === LAND) {
      grid[i][j] = WATER;
      flood(i + 1, j);
      flood(i - 1, j);
      flood(i, j + 1);
      flood(i, j - 1);
    }
  };

  let islandCount = 0;
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === LAND) {
        islandCount++;
        flood(i, j);
      }
    }
  }

  return islandCount;
};
// @lc code=end
